### PR TITLE
Subaccount module part2

### PR DESCRIPTION
**Pull request by faultytolly**
## Description

---

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Added Features

- [ ]

---

## Modified Features

- [ ]

---

## Test Cases

- [ ] Test A
- [ ] Test B

---

## Change Log

---

## Documentation

---

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
---

### CODE REVIEW

File: x/subaccount/keeper/create_subaccount.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+	senderAccount, _ := sdk.AccAddressFromBech32(request.Sender)
+	subaccountOwner, _ := sdk.AccAddressFromBech32(request.SubAccountOwner)
+	if m.keeper.HasSubAccount(sdkContext, subaccountOwner) {
+		return nil, types.ErrSubaccountAlreadyExist
+	}
+
+	subaccountID := m.keeper.NextID(sdkContext)
+
+	// ALERT: If someone frontruns the account creation, will be overwritten here
+	subaccountAddress := types.NewAddressFromSubaccount(subaccountID)
```
Comment (scorpioborn) [inline x/subaccount/keeper/create_subaccount.go:40]: I'm wondering, Is creating module accounts is a good approach to store sub accounts? Can we use BasicAccounts for this purpose?

File: x/subaccount/keeper/create_subaccount.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	address := m.accountKeeper.NewAccountWithAddress(sdkContext, subaccountAddress)
+	m.accountKeeper.SetAccount(sdkContext, address)
+
+	denom := m.keeper.GetParams(sdkContext).LockedBalanceDenom
+	err = m.bankKeeper.SendCoins(sdkContext, senderAccount, subaccountAddress, sdk.NewCoins(sdk.NewCoin(denom, moneyToSend)))
```
Comment (scorpioborn) [inline x/subaccount/keeper/create_subaccount.go:45]: Formerly, `OrderBook` module was responsible for transferring the tokens between accounts and module accounts using `funder` inderface. @3eyedraga Do you think that `subaccount` module should use them as well?

**Review by scorpioborn — CHANGES_REQUESTED** [review]

File: x/subaccount/keeper/create_subaccount.go (hunk 1 above)
Comment (faultytolly) [reply x/subaccount/keeper/create_subaccount.go:40]: These accounts are BaseAccounts, not module accounts and this is needed for some module account trickery happening in the SDK, please refer to how cosmwasm does contract creation, it's basically for the same purpose of ours: https://github.com/CosmWasm/wasmd/blob/main/x/wasm/keeper/keeper.go#L280

File: x/subaccount/keeper/create_subaccount.go (hunk 2 above)
Comment (faultytolly) [reply x/subaccount/keeper/create_subaccount.go:45]: As mentioned https://github.com/sge-network/sge/pull/208/files#r1259558460, this cannot be a module account, it's a normal base account.

File: proto/sge/subaccount/params.proto [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+import "gogoproto/gogo.proto";
+
+option go_package = "github.com/sge-network/sge/x/subaccount/types";
+
+// Params defines the parameters for the module.
+message Params {
+  option (gogoproto.goproto_stringer) = false;
+
+  // balance_denom defines the denom of the locked balances.
+  string locked_balance_denom = 1;
```
[the remark this answers is not in the document]
Comment (faultytolly) [reply proto/sge/subaccount/params.proto:13]: This should still be parametrised to allow spawning testnets where the token can be != `sge`.

File: x/subaccount/keeper/create_subaccount.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	sdkContext := sdk.UnwrapSDKContext(ctx)
+	moneyToSend := sdk.NewInt(0)
+
+	err := request.ValidateBasic()
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid request")
+	}
+
+	for _, balanceUnlock := range request.LockedBalances {
+		if balanceUnlock.UnlockTime.Unix() < sdkContext.BlockTime().Unix() {
```
[the remark this answers is not in the document]
Comment (faultytolly) [reply x/subaccount/keeper/create_subaccount.go:24]: It's generally not a good idea to implement methods on protobuf messages:

`defining interfaces on protobuf types moves implementation into APIs - for achieving proper semantic versioning (https://github.com/cosmos/cosmos-sdk/discussions/10162) we want to move generated proto files to separate go modules, but coupling generated types with implementation details in interfaces complicates this`

Ref: https://github.com/cosmos/cosmos-sdk/discussions/10368

**Review by 3eyedraga — APPROVED** [review]

**Pull request merged**